src/xdg-autostart-generator/xdg-autostart-generator.c

```diff
@@ -0,0 +1,771 @@
+                        if (r < 0) {
+                                log_warning_errno(r, "failed to parse %s, ignoring it", service->path);
+                                service->hidden = 1;
+                        }
+
+                        r = hashmap_put(all_services, service->name, service);
+                        if (r < 0)
+                                return log_oom();
+                        service = NULL;
+                }
+        }
+
+        return 0;
+}
+
+static int format_exec_start(
+                const char *exec,
+                char **exec_start_out) {
+
+        _cleanup_strv_free_ char **exec_split = NULL;
+        _cleanup_free_ char *exec_start = NULL;
+        size_t allocated = 0, length = 0;
+        size_t n;
+        bool first_arg;
+        char **arg;
+        int r;
+
+        /*
+         * Unfortunately, there is a mismatch between systemd's idea of $PATH
+         * and XDGs. i.e. we need to ensure that we have an absolute path to
+         * support cases where $PATH has been modified from the default set.
+         *
+         * Note that this is only needed for development environments though;
+         * so while it is important, this should have no effect in production
+         * environments.
+         *
+         * To be compliant with the XDG specification, we also need to strip
+         * certain parameters and such. Doing so properly makes parsing the
+         * command line unavoidable.
+         *
+         * NOTE: Technically, XDG only specifies " as quotes, while this also
+         *       accepts '.
+         */
+        exec_split = strv_split_full(exec, WHITESPACE, SPLIT_QUOTES | SPLIT_RELAX);
+        if (!exec_split)
+                return -ENOMEM;
+
+        /*
+         * Remove any standardised XDG fields; we assume they never appear as
+         * part of another argument as that just does not make any sense as
+         * they can be empty (GLib will e.g. turn "%f" into an empty argument).
+         * Other implementations may handle this differently.
+         */
+        strv_remove(exec_split, "%f"); strv_remove(exec_split, "%F");
+        strv_remove(exec_split, "%u"); strv_remove(exec_split, "%U");
+        strv_remove(exec_split, "%d"); strv_remove(exec_split, "%D");
+        strv_remove(exec_split, "%n"); strv_remove(exec_split, "%N");
+        strv_remove(exec_split, "%i"); /* Location of icon, could be implemented. */
+        strv_remove(exec_split, "%c"); /* Translated application name, could be implemented*/
+        strv_remove(exec_split, "%k"); /* Location of desktop file, could be implemented. */
+        strv_remove(exec_split, "%v");
+        strv_remove(exec_split, "%m");
+
+        if (strv_length (exec_split) == 0) {
+                log_info("Exec line is empty");
+                return -EINVAL;
+        }
+
+        first_arg = 1;
+        STRV_FOREACH(arg, exec_split) {
```

Comment:
Let's restructure this:
```C
unsigned i, n;
exec_split = strv_split_full(...);

for (i = n = 0; exec_split[i]; i++) {
         if (STR_IN_SET(exec_split[i],
                                    "%f", "%F",
                                     "%u, "%u",
                                     ...))
                      continue;

          _cleanup_free_ char *new_value = NULL;

          /* unescape and do other processing */

          free_and_replace(exec_split[n++], new_value);
}
while (n <= i)
      exec_split[n++] = mfree(exec_split[n++]);
char *t = strv_join(exec_split, " ");
if (!t)
     ...;
*ret_exec_start = t;
```
This avoid any quadratic loops over the arguments. Normally this shouldn't be a problem, but we shouldn't go belly-up if somebody has a file with one thousand %f's.

Reply:
Yeah, except that some of those things are a bad idea ;-)

`free_and_replace` expands the first argument twice for example. But got a solution.